packages/valory/skills/abstract_round_abci/tests/test_base.py

```diff
@@ -1790,6 +1912,99 @@ def test_update_round(
+        self.round_sequence._blockchain.add_block(block)
+
+        with mock.patch.object(
+            self.round_sequence.current_round,
+            "end_block",
+            return_value=current_round_result,
+        ), mock.patch.object(
+            self.round_sequence.background_round,
+            "end_block",
+            return_value=background_round_result,
+        ):
+            self.round_sequence._update_round()
+
+        if background_round_result is None and current_round_result is None:
+            assert (
+                self.round_sequence._last_round_transition_timestamp
+                != self.round_sequence._blockchain.last_block.timestamp
+            )
+            assert (
+                self.round_sequence._last_round_transition_height
+                != self.round_sequence._blockchain.height
+            )
+            assert (
+                self.round_sequence._last_round_transition_root_hash
+                != self.round_sequence.root_hash
+            )
+            assert (
+                self.round_sequence._last_round_transition_tm_height
+                != self.round_sequence.tm_height
+            )
+            process_event_mock.assert_not_called()
+        elif background_round_result is None and current_round_result is not None:
+            assert (
+                self.round_sequence._last_round_transition_timestamp
+                == self.round_sequence._blockchain.last_block.timestamp
+            )
+            assert (
+                self.round_sequence._last_round_transition_height
+                == self.round_sequence._blockchain.height
+            )
+            assert (
+                self.round_sequence._last_round_transition_root_hash
+                == self.round_sequence.root_hash
+            )
+            assert (
+                self.round_sequence._last_round_transition_tm_height
+                == self.round_sequence.tm_height
+            )
+            process_event_mock.assert_called_with(
+                current_round_result[-1],
+                result=current_round_result[0],
+            )
+        elif background_round_result is not None:
+            assert (
+                self.round_sequence._last_round_transition_timestamp
+                == self.round_sequence._blockchain.last_block.timestamp
+            )
+            assert (
+                self.round_sequence._last_round_transition_height
+                == self.round_sequence._blockchain.height
+            )
+            assert (
+                self.round_sequence._last_round_transition_root_hash
+                == self.round_sequence.root_hash
+            )
+            assert (
+                self.round_sequence._last_round_transition_tm_height
+                == self.round_sequence.tm_height
+            )
+            process_event_mock.assert_called_with(
+                background_round_result[-1],
+                result=background_round_result[0],
+            )
```

Comment:
what I dislike about the current implementation is that there is no `else` clause. 

```
same_timestamp = self.round_sequence._last_round_transition_timestamp
                != self.round_sequence._blockchain.last_block.timestamp
same_height =  self.round_sequence._last_round_transition_height
                != self.round_sequence._blockchain.height
same_root_hash = self.round_sequence._last_round_transition_root_hash
                != self.round_sequence.root_hash
same_tm_height = self.round_sequence._last_round_transition_tm_height
                != self.round_sequence.tm_height
                
conditions_are_equal = (same_timestamp, same_height, same_root_hash, same_tm_height)
                
if background_round_result is current_round_result is None:
    assert not any(conditions_are_equal)
    process_event_mock.assert_not_called()
else:  # ?
     assert all(conditions_are_equal)
     process_event_mock.assert_called_with(
         current_round_result[-1],
         result=current_round_result[0],
     )
```

Reply:
The case when `background_round_result is None` is covered already in `test_update_round`.

Reply:
But I understand what you mean. We could refactor this and have it all in one method.

Reply:
even if it cannot happen, then I'd rather have an `else: raise` so another dev knows that something won't ever silently pass